File: LeetCode/Medium/0095_unique-binary-search-trees-ii/20240626-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


dp = {}


def f(l: int, r: int) -> list[TreeNode | None]:
    if l >= r:
        return [None]

    if (l, r) in dp:
        return dp[l, r]

    ans = []
    for i in range(l, r):
        left_nodes = f(l, i)
        right_nodes = f(i + 1, r)
        for left_node in left_nodes:
            for right_node in right_nodes:
                ans.append(TreeNode(i, left_node, right_node))

    dp[l, r] = ans
    return ans


class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        return f(1, n + 1)

File: LeetCode/Easy/0094_binary-tree-inorder-traversal/20240626-1.rs
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut ans = vec![];
        f(root.as_ref(), &mut ans);
        ans
    }
}

fn f(node: Option<&Rc<RefCell<TreeNode>>>, ans: &mut Vec<i32>) {
    if node.is_none() {
        return;
    }

    let node = node.unwrap().borrow();
    f(node.left.as_ref(), ans);
    ans.push(node.val);
    f(node.right.as_ref(), ans);
}

struct Solution;

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
